tests/unit/source_code/test_dependencies.py

```diff
@@ -0,0 +1,238 @@
+from typing import BinaryIO
+from unittest.mock import create_autospec
+
+import pytest
+from databricks.sdk import WorkspaceClient
+from databricks.sdk.service.workspace import ObjectInfo, Language, ObjectType
+
+from databricks.labs.ucx.hive_metastore.migration_status import MigrationIndex
+from databricks.labs.ucx.source_code.dependencies import DependencyLoader
+from databricks.labs.ucx.source_code.languages import Languages
+from databricks.labs.ucx.source_code.notebook import Notebook
+from databricks.labs.ucx.source_code.source_migrator import SourceCodeMigrator
+from tests.unit import _load_sources
+
+
+def test_build_dependency_graph_visits_notebook_notebook_dependencies():
+    paths = ["root3.run.py.txt", "root1.run.py.txt", "leaf1.py.txt", "leaf2.py.txt"]
+    sources: dict[str, str] = dict(zip(paths, _load_sources(Notebook, *paths)))
+    visited: dict[str, bool] = {}
+
+    # can't remove **kwargs because it receives format=xxx
+    # pylint: disable=unused-argument
+    def download_side_effect(*args, **kwargs):
+        filename = args[0]
+        if filename.startswith('./'):
+            filename = filename[2:]
+        visited[filename] = True
+        result = create_autospec(BinaryIO)
+        result.__enter__.return_value.read.return_value = sources[filename].encode("utf-8")
+        return result
+
+    def get_status_side_effect(*args):
+        path = args[0]
+        return ObjectInfo(path=path, language=Language.PYTHON, object_type=ObjectType.NOTEBOOK)
```

Comment:
i see these download side effects repeated. don't you want to use `workspace_client_mock`, that reads real files from the real folder?.. you can mock the get_status behavior based on file extensions. Would be easier, given the scale we aim for.
https://github.com/databrickslabs/ucx/blob/main/tests/unit/__init__.py#L134-L152

Reply:
I'd love to not repeat this code but `workspace_client_mock` doesn't implement `download`. Plus I need to map the import name (without extension) to a real name (with 2 extensions). The above code is inspired by similar code in `test_notebook_migrator.py`. I think once the resolver is fully implemented it will be a good time to simplify these tests ? See https://github.com/databrickslabs/ucx/issues/1369